### PR TITLE
chore(actors): refactor allgather

### PR DESCRIPTION
**Pull request by stoksc**
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/task/allgather/allgather.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	defer g.mu.Unlock()
+
+	if _, ok := g.watchers[msg.WatcherID]; ok {
+		// If this peer has already connected, just respond with the watcher again. This is only
+		// possible if it disconnects and reconnects since the original actor ask blocks forever.
+		return Watcher{C: g.watchers[msg.WatcherID]}
+	}
+
+	// Channel is size 1 since data info will only ever be sent once and we'd rather not block.
+	w := make(chan Result, 1)
```
Comment (corban-beaird) [inline master/internal/task/allgather/allgather.go:89]: praise: setting the channel size to 1 is a really clean way to minimize resource usage!

**Review by corban-beaird — APPROVED** [review]
Looks great!

**Comment by stoksc** [comment]
should've marked this draft, have some changing incoming, including the tests

**Comment by stoksc** [comment]
closing in favor of an upstream branch

**Pull request closed without merging**